### PR TITLE
Add ERC20MintableBurnable test

### PR DESCRIPTION
**Pull request by yudetamago**
Translate `erc20_test_1.py` ( https://github.com/ethereum/vyper/blob/97ed5fec89947d41fd91f5190c4478ffcadbb9fb/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py ) to pytest

* convert
    * `deposit` to `mint`
    * `withdraw` to `burn`
* remove checking wei balance like `self.assertEqual(initial_a1_balance - self.s.head_state.get_balance(self.t.a1), 2)` because `mint` and `burn` are irrelevant to wei
* add tests about `msg.sender == self.minter` and to `_to != ZERO_ADDRESS`  (L58-64)
* remove `test_payability` because there are no payable functions in ERC20MintableBurnable
* remove emitting no event assertions in `test_raw_logs` because checking events with call operation or failed tx may not be supported in eth_tester?
* remove tests about overflow (`test_maxInts`, `test_bad_transfer`, `test_bad_deposit`, and `test_bad_transferFrom`) because overflow assertion is supported by default in vyper
* remove `test_failed_send_in_withdraw` because it tests solidity code?

### CODE REVIEW

File: tests/erc20/test_erc20_mintable_burnable.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+    with open('../contracts/erc20/ERC20MintableBurnable.vy') as f:
+        code = f.read()
+    name = bytes_helper("name", 32)
+    symbol = bytes_helper("sym", 32)
+    c = get_contract(code, name, symbol, 0, 0)
+    return c
+
+
+@pytest.fixture
+def get_log_args(get_logs):
```
Comment (nrryuya) [inline tests/erc20/test_erc20_mintable_burnable.py]: It would be nice to put this in conftest.py for reuse?

**Comment by nrryuya** [comment]
>remove test_failed_send_in_withdraw because it tests solidity code?

I think the solidity contract is there to call the Vyper contract as an external contract.

File: tests/erc20/test_erc20_mintable_burnable.py (hunk 1 above)
Comment (yudetamago) [reply tests/erc20/test_erc20_mintable_burnable.py]: It seems so, but conftest.py is coped from https://github.com/ethereum/vyper/blob/97ed5fec89947d41fd91f5190c4478ffcadbb9fb/tests/conftest.py
To do PR to vyper repo, it might be necessary to change vyper repo's one.

**Comment by yudetamago** [comment]
>I think the solidity contract is there to call the Vyper contract as an external contract.

ah, I got it.

**Comment by nrryuya** [comment]
>remove emitting no event assertions in test_raw_logs because checking events with call operation or failed tx may not be supported in py-evm?

py-evm is not used in our repo, is it?
I think event is not emitted in revert?

**Comment by yudetamago** [comment]
> py-evm is not used in our repo, is it?

Ah, I misunderstood, eth_tester is used. (EDITED)

> I think event is not emitted in revert?

Right, the original tests check it.
https://github.com/ethereum/vyper/blob/97ed5fec89947d41fd91f5190c4478ffcadbb9fb/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py#L272-L280

File: tests/erc20/test_erc20_mintable_burnable.py (hunk 1 above)
Comment (nrryuya) [reply tests/erc20/test_erc20_mintable_burnable.py]: Be animal

**Comment by nrryuya** [comment]
>remove tests about overflow (test_maxInts, test_bad_transfer, test_bad_deposit, and test_bad_transferFrom) because overflow assertion is supported by default in vyper

I think they are necessary to check all the corner cases. If we have a variable X whose range is from A to B, we need to test with A and B for "complete testing". The statement that the tx fail in overflows is  the implicit specification so we need to check it.

BTW,tThe issue below was fixed in Vyper compiler so we can mint MAX_UINT256.
https://github.com/ethereum/vyper/blob/97ed5fec89947d41fd91f5190c4478ffcadbb9fb/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py#L175

**Comment by nrryuya** [comment]
>The statement that the tx fail in overflows is the implicit specification so we need to check it.

This isn't limited to overflow because any type has range. e.g. This code causes revert.

```
    a: int128 = MAX_INT128 - 1
    a += 2
```

File: tests/erc20/test_erc20_mintable_burnable.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+ZERO_ADDRESS = '0x0000000000000000000000000000000000000000'
+
+
+@pytest.fixture
+def c(get_contract, w3, bytes_helper):
+    with open('../contracts/erc20/ERC20MintableBurnable.vy') as f:
+        code = f.read()
+    name = bytes_helper("name", 32)
+    symbol = bytes_helper("sym", 32)
```
Comment (nrryuya) [inline tests/erc20/test_erc20_mintable_burnable.py]: symbols are CAPITALIZED basically?

Comment (nrryuya) [reply tests/erc20/test_erc20_mintable_burnable.py]: How about `Vypercoin`, `VYP` ? :)

Comment (yudetamago) [reply tests/erc20/test_erc20_mintable_burnable.py]: Niiiice!

Comment (yudetamago) [reply tests/erc20/test_erc20_mintable_burnable.py]: https://github.com/LayerXcom/verified-vyper-contracts/pull/50

File: tests/erc20/test_erc20_mintable_burnable.py [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+    assert args._value == 1
+
+    args = get_log_args(c.transferFrom(
+        a2, a3, 0, transact={'from': a1}), c, 'Transfer')
+    assert args._from == a2
+    assert args._to == a3
+    assert args._value == 0
+
+
+def test_failed_send_in_withdraw(c, w3):
```
Comment (nrryuya) [inline tests/erc20/test_erc20_mintable_burnable.py]: I think we don't need this test because it tests `withdraw` function on fallback in sending ether.

Comment (yudetamago) [reply tests/erc20/test_erc20_mintable_burnable.py]: :agree:
(Why did I add that? 🤔 )

Comment (nrryuya) [reply tests/erc20/test_erc20_mintable_burnable.py]: You are deceived by me.

**Comment by nrryuya** [comment]
I don't know why it's in the original but test for Solidity ERC20 can be omitted?
https://github.com/ethereum/vyper/blob/97ed5fec89947d41fd91f5190c4478ffcadbb9fb/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py#L393

File: tests/erc20/test_erc20_mintable_burnable.py (hunk 3 above)
Comment (yudetamago) [reply tests/erc20/test_erc20_mintable_burnable.py]: 😂

**Comment by nrryuya** [comment]
All what remains are the "bad_code" test. I'll work on this.

File: tests/erc20/test_erc20_mintable_burnable.py (hunk 3 above)
Comment (yudetamago) [reply tests/erc20/test_erc20_mintable_burnable.py]: https://github.com/LayerXcom/verified-vyper-contracts/pull/53

**Comment by yudetamago** [comment]
yeah!
